### DISCUSSION TITLE
Issue With ZodLazy Losing Type Safety

### DISCUSSION THREAD

**Discussion by tdfranklin** [colinhacks/zod#3027] [Q&A] [answered]
Hi! Having an issue I hope you can help with (I'm still pretty new to using Zod).

So, to backtrack slightly, I first had an issue with what I believe was a circular dependency where schema's referenced in two separate files reference each other:

```typescript
// shipment.ts
import { z } from 'zod'
import { orderSchema } from './order'

export const shipmentSchema = z.object({
  id: z.number(),
  orders: z.array(orderSchema),
  // ...etc
})

// order.ts
import { z } from 'zod'
import { shipmentSchema } from './shipment'

export const orderSchema = z.object({
  id: z.number(),
  name: z.string(),
  // ...etc
})

export const orderExtendedSchema = orderSchema.extend({
  shipment: shipmentSchema,
  // ...etc
})
```

Adding this code immediately broke the app, displaying a blank page with a console error of `Uncaught ReferenceError: Cannot access 'orderSchema' before initialization`

So after some investigation, it seemed like z.lazy might come to the rescue, so I updated shipmentSchema.orders to instead be `z.array(z.lazy(() => orderSchema))` which solved the error and everything seemed to work, except that now, I seem to lose the typing for orders when I call the function that parses (let me explain more below with another contrived example):

```typescript
// shipment-template.tsx
import { getShipment } from './shipment'

const shipments = await getShipments(123)
shipments.orders // <= this is typed as any[]

// shipment.ts
import { z } from 'zod'
import { api } from './api'
import { orderSchema } from './order'

async function getShipment(id: number) {
  try {
    const response = await api.get(`shipments/${id}`)
    const shipment = shipmentSchema.parse(response.data)
    shipment.orders // <= this is typed correctly as orderSchema

    return shipment
  } catch (e) {
    console.error(e)
  }
}

export const shipmentSchema = z.object({
  id: z.number(),
  orders: z.array(z.lazy(() => orderSchema)),
  // ...etc
}) 

// order.ts
import { z } from 'zod'
import { shipmentSchema } from './shipment'

export const orderSchema = z.object({
  id: z.number(),
  name: z.string(),
  // ...etc
})

export const orderExtendedSchema = orderSchema.extend({
  shipment: shipmentSchema,
  // ...etc
})
```

So in `getShipment` API call, after parsing the response, `shipment.orders` is typed correctly. However, whenever I call `getShipment` in my application, suddenly `shipment.orders` is typed as `any[]`. Any idea what is causing this or how to fix it? Is there a better way I should be handling the circular dependency instead of using `z.lazy()`?

**Comment by JacobWeisenburger** [comment]
perhaps this would help: https://github.com/colinhacks/zod#recursive-types

**Reply by tdfranklin** [reply, marked as answer]
Thanks for the suggestion! I was playing around with that but couldn't get it to work. Turns out, the issue was with how I was importing the schema's. Once I swapped to using relative paths instead of absolute, it started working!
